crates/core/tedge_api/goals.md

```diff
@@ -0,0 +1,363 @@
+
+ThinEdge can thus be understood as being a common-core of a message passing router
+with a collection of plugins that ultimately define what it actually _does_.
+
+Plugins come in two forms:
+
+- Run-time & external, which are provided to Thin-Edge and meant to
+  interoperate with it specifically through stdin/stdout or some other
+  well-specified interface (cf. external plugins further down)
+    - These are for example custom built executables that a user provides and
+      wishes to not integrate into ThinEdge for various reasons (e.g. language
+      differences or license)
+- Compile-time & built-in, which are compiled into the actual ThinEdge binary
+  and written in Rust.
+    - These offer the advantage of simplifying deployment as well as assurances
+      w.r.t. the messages (i.e. changes)
+
+Both ways are _functionally equivalent_ and _support the same features_.
+
+We aim to provide plugins for all common server and use-cases, see further down
+if you wish to extend Thin-Edge yourself.
+
+## Configuring your plugins
+
+ThinEdge is on its own merely a collection of plugin kinds, and requires a
+configuration to do any useful work. A common configuration includes a cloud
+mapper, some device management plugins as well as some data sources.
+
+> One core idea is that you can create as many plugin instances of a single
+> plugin _kind_ as you wish. For example you could have multiple "File Watcher"
+> instances that each watch a different file. 
+
+
+Here is an fictuous configuration that would connect the device to the acme cloud.
+
+```toml
+[plugins.acme]
+kind = "acme_mapper"
+[plugins.acme.configuration]
+tenant = "coyote"
+software_mgmt_plugin = "pacman" # Here we tell the azure_mapper plugin where to
+                                # send software management requests
+service_mgmt_plugin = "systemd"
+
+[plugins.pacman]
+kind = "pacman_handler"
+[plugins.pacman.configuration]
+allow_destructive_operations = true # Per default the plugin does not allow
+                                    # adding/removing packages
+
+[plugins.systemd]
+kind = "systemd_handler"
+configuration = { 
+    restrict_units_to = ["nginx.service"]  # Only allow interacting with the nginx service
+}
+```
+
+Of note here is that everying in the `plugins.<plugin id>.configuration` space
+is per plugin! Each plugin exposes its own set of configurations depending on
+its needs and abilities. Nonetheless some parts are probably more common:
+
+- Per default plugins should strive to do the 'safe' thing. Ambiguitiy should
+  be reduced as much as possible and if defaults are unclear should either
+  force the user to specify it or do an idempotent safe/pure operation. (e.g.
+  the pacman plugin only allows listing packages per default)
+- Plugins don't guess where to send their messages to. If the `acme_mapper`
+  receives a 'Restart "nginx" service' message it needs to be configured to
+  tell it where to find the destination for it.
```

Comment:
> - Plugins don't guess where to send their messages to.

I have a preference for this compared some discovery mechanism. This remove ambiguities and makes possible to existence of several instance of the same plugin kind. e.g. an MQTT plugin instance sending measurements to the cloud and another to some local analytics engine.

> If the `acme_mapper`
  receives a 'Restart "nginx" service' message it needs to be configured to
  tell it where to find the destination for it.

Here, I wonder what is the right level to route messages? The example sounds too fine grains.